### PR TITLE
Get repo visibility from GitHub

### PR DESCRIPTION
**Pull request by ddgenome**
Use the GitHub API to determine repo visibility.  Remove org argument
to functions, derive it using GitHubRepo.  RestartBuildFunction still
takes a parameter indicating which Travis API to hit since it does not
get repo owner and name but a build ID, so there is no way to derive
the Travis API.  We did change the parameter to accept "public" and
"private".

Move retry logic out into own object so it can be used by both Travis
and GitHub methods.

Update rug to 1.0.0-m.3.

Make RepoSlug type because the owner and name were almost always used
together and with a type we can avoid putting multiple string
arguments in the wrong order.

Make the GitHubRepo trait simpler, preferring to pass in arguments to
functions rather than storing state in the object.

Make types for tokens so they do not get mixed up

This is working toward atomist/travis-rugs#8

### CODE REVIEW

File: src/main/scala/com/atomist/rug/functions/travis/GitHubRepo.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  /** Determine correct Travis API endpoint for a GitHub.com repository.
+    *
+    * @param repoSlug repostory "owner/name"
+    * @param token GitHub personal access token with access to repoSlug
+    * @return TravisAPIEndpoint to use with repoSlug
+    */
+  def travisEndpoint(repoSlug: RepoSlug, token: GitHubToken): TravisAPIEndpoint = {
+    val repo = getRepo(repoSlug, token)
+    val isPrivate = repo.get("private").asInstanceOf[Boolean]
+    if (isPrivate) TravisComEndpoint
```
Comment (kipz) [inline src/main/scala/com/atomist/rug/functions/travis/GitHubRepo.scala:27]: Should we use curly braces here?

**Comment by kipz** [comment]
LGTM - just minor point around style.

File: src/main/scala/com/atomist/rug/functions/travis/GitHubRepo.scala (hunk 1 above)
Comment (ddgenome) [reply src/main/scala/com/atomist/rug/functions/travis/GitHubRepo.scala:27]: You mean for the `if`? I was actually thinking of making it one line.

```scala
if (isPrivate) TravisComEndpoint else TravisOrgEndpoint
```

I know TypeScript/TSLint prefers the curly braces, but Scala does not.

Comment (kipz) [reply src/main/scala/com/atomist/rug/functions/travis/GitHubRepo.scala:27]: IMO, we should _always_ use curly braces in Java, JS, TS, Scala etc. But I guess we haven't agreed this yet. :)

**Pull request merged**